src/opm/output/eclipse/WriteRFT.cpp

```diff
@@ -1010,13 +1013,326 @@ namespace {
+        rftFile.write("SEGBRNO", this->branch_id_);
+        rftFile.write("SEGNXT", this->neighbour_id_);
+
+        rftFile.write("BRNST", this->branch_start_segment_);
+        rftFile.write("BRNEN", this->branch_end_segment_);
+    }
+
+    void SegmentRecord::collectRecordData(const ::Opm::UnitSystem& usys,
+                                          const ::Opm::Well&       well,
+                                          const ::Opm::data::Well& wellSol)
+    {
+        const auto& segments = well.getSegments();
+        const auto& xseg = wellSol.segments;
+
+        for (auto begin = segments.begin(), end = segments.end();
+             begin != end; ++begin)
+        {
+            auto segSolPos = xseg.find(begin->segmentNumber());
+            if (segSolPos == xseg.end()) {
+                continue;
+            }
+
+            this->addSegment(usys, segments, begin, segSolPos->second);
+        }
+
+        if (this->nSeg() > std::size_t{0}) {
+            this->defineBranches(segments);
+        }
+    }
+
+    void SegmentRecord::defineBranches(const ::Opm::WellSegments& wellSegs)
+    {
+        auto branchSegments = CSRIndexRelation {};
+        const auto minBranchID = 1;
+
+        branchSegments.build(wellSegs.size(), minBranchID,
+            [&wellSegs](const int ix) { return wellSegs[ix].branchNumber(); },
+            OrderSegments { wellSegs });
+
+        const auto maxBranchID = branchSegments.maxBinId();
+        this->branch_start_segment_.assign(maxBranchID, 0);
+        this->branch_end_segment_  .assign(maxBranchID, 0);
+
+        auto segNum = [&wellSegs](const int segIx) {
+            return wellSegs[segIx].segmentNumber();
+        };
+
+        for (auto branch = minBranchID; branch <= maxBranchID; ++branch) {
+            auto [begin, end] = branchSegments.bin(branch);
+            if (end == begin) {
+                // Empty branch (no segments).  Unexpected.
+                continue;
+            }
+
+            const auto branchIx = branch - minBranchID;
+            this->branch_start_segment_[branchIx] = segNum(*begin);
+            this->branch_end_segment_  [branchIx] = segNum(*(end - 1));
+        }
+    }
+
+    void SegmentRecord::addSegment(const ::Opm::UnitSystem&   usys,
+                                   const ::Opm::WellSegments& segments,
+                                   SegmentPos                 segPos,
+                                   const Opm::data::Segment&  segSol)
+    {
+        this->recordPhysicalLocation(usys, segments, segPos);
+        this->recordSegmentConnectivity(*segPos);
+        this->recordSegmentProperties(usys, *segPos);
+        this->recordDynamicState(usys, *segPos, segSol);
+    }
+
+    void SegmentRecord::recordPhysicalLocation(const ::Opm::UnitSystem&   usys,
+                                               const ::Opm::WellSegments& segments,
+                                               SegmentPos                 segPos)
```

Comment:
i don't see any reason why this takes an iterator? i may be missing something, but taking a const ref here like in the other methods would allow addSegment to take a const ref, which would allow a range based for in collectRecordData.

Reply:
> I'll switch to using a range-for.

I've pushed an update to this effect.